src/appleseed/foundation/math/sampling/qmcsamplingcontext.h

```diff
@@ -279,7 +333,7 @@ template <typename RNG>
 template <typename T>
 inline T QMCSamplingContext<RNG>::next2()
 {
-    return next2(Tag<T>());
+    return next2(TypeTag<T>());
```

Comment:
My initial idea was to have one "generic" next2 that is used most of the time and some "specific" next2 such as RNG or CMJ. This can be achieved by having one "generic" ```T next2(tag<T, Method>)``` that will consequently call Halton and three specific:
```Vector<T, N> next2(tag<Vector<T, N>, M_Halton>)```
```Vector<T, 2> next2(tag<Vector<T, 2>, M_CMJ>)``` and 
```T next2(tag<T, M_RNG>)```; 
When you pass CMJ with improper signature it fallback to generic next2 automagically since there is no specialization for it; 
So you basically could check if CMJ is enabled and call next2 with M_CMJ right here no matter if the signature is ok;

```inline T QMCSamplingContext<RNG>::next2(Tag<T, Method>)``` should only (dumb) dispatch execution flow to (as it does now)
```inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>, Method>)``` which dispatches any calls to 
```inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>, M_Halton>)```

So, there is how it is supposed to work:
1) Call ```next2(<double>)``` -> [if CMJ is enabled] ```next2(<tag<double, M_CMJ>>)``` -> ```(next2<tag<Vector<double, N>, CMJ>>)``` -> [there is no such specialization, fallback to generic]   ```next2(<tag<Vector<double, N>, Method>>)``` -> [Generic will dispatch to Halton]```next2(<tag<Vector<double, N>, M_Halton>>)```
2)  Call ```next2(<double>)``` -> [if CMJ is disabled] ```next2(<tag<double, M_Halton>>)``` -> [there is specialization for Halton, use it] ```next2(<tag<Vector<double, N>, M_Halton>>)```
3)  Call ```next2(<double>)``` -> [mode is RNG] ```next2(<tag<double, M_RNG>>)``` -> [here we go, specialization for RNG] ```next2(<tag<Vector<double, N>, M_RNG>>)```
4) Call ```next2(<Vector2d>) ``` -> [if CMJ is enabled, there is specialization] ```next2(<tag<Vector<double, 2>, M_CMJ>>)```
5) Call ```next2(<Vector2d>) ``` -> [if CMJ is disabled, there is specialization] ```next2(<tag<Vector<double, 2>, M_Halton>>)```
e.t.c

You can notice there is only 1 tag instead of 2; The whole number of functions is 6:
```cpp 
T next2(T) // we set method here
T next2(tag<T, Method>) // here we ONLY dispatch tag<T, already_defined_method> to tag<Vector<T, 1>, already_defined_method>```
Vector<T, N> next2(tag<Vector<T, N>, Method>) // generic, this simply dispatch to Halton
Vector<T, N> next2(tag<Vector<T, N>, M_Halton>)
Vector<T, N> next2(tag<Vector<T, N>, M_RNG>)
Vector<T, 2> next2(tag<Vector<T, 2>, M_CMJ>)
```

Reply:
Another good point of my approach is that you can simply add new methods. Just implement ```cppVector<T, 1> next2(<Vector<T, 1>, M_CMJ>)``` and you get 1d cmj with no hustle:)